From c7bd3fefdfe39c7f0f817fb18ff06ea1ff45a744 Mon Sep 17 00:00:00 2001
From: Tiejun Chen <tiejun.chen@windriver.com>
Date: Fri, 28 Aug 2009 16:36:32 +0800
Subject: [PATCH 28/38] powerpc/83xx: set system i/o pins when enable eSDHC controller

This is extracted from patch
kernel-2.6.25-MPC837xE-RDB-add-esdhc-support.patch in rev 20081113 board
support ISO image for MPC837xE RDB.

The eSDHC controller SD_DAT1/2/3 are multiplexed with other device.
We shoud set these pins to the correct function before using the eSDHC
controller.

Integrated-by: Kevin Hao <kexin.hao@windriver.com>
---
 arch/powerpc/platforms/83xx/mpc837x_rdb.c |   27 +++++++++++++++++++++++++++
 arch/powerpc/platforms/83xx/mpc83xx.h     |    4 ++++
 2 files changed, 31 insertions(+), 0 deletions(-)

diff --git a/arch/powerpc/platforms/83xx/mpc837x_rdb.c b/arch/powerpc/platforms/83xx/mpc837x_rdb.c
index da030af..55f9748 100644
--- a/arch/powerpc/platforms/83xx/mpc837x_rdb.c
+++ b/arch/powerpc/platforms/83xx/mpc837x_rdb.c
@@ -18,9 +18,32 @@
 #include <asm/ipic.h>
 #include <asm/udbg.h>
 #include <sysdev/fsl_pci.h>
+#include <sysdev/fsl_soc.h>
 
 #include "mpc83xx.h"
 
+#ifdef CONFIG_MMC_SDHCI_OF
+static int mpc837xrdb_sd_cfg(void)
+{
+	void __iomem *immap = NULL;
+	unsigned long sicrl, sicrh;
+
+	/* Configure the multiplexing of pins */
+	immap = ioremap(get_immrbase(), 0x1000);
+	if (!immap)
+		return -ENOMEM;
+
+	sicrl = in_be32(immap + MPC83XX_SICRL_OFFS) & ~MPC837X_SICRL_USBB_MASK;
+	sicrh = in_be32(immap + MPC83XX_SICRH_OFFS) & ~MPC837X_SICRH_SPI_MASK;
+
+	out_be32(immap + MPC83XX_SICRL_OFFS, (sicrl | MPC837X_SICRL_SD));
+	out_be32(immap + MPC83XX_SICRH_OFFS, (sicrh | MPC837X_SICRH_SD));
+	udelay(100);
+	iounmap(immap);
+	return 0;
+}
+#endif
+
 extern int mpc837x_usb_cfg(void);
 
 /* ************************************************************************
@@ -42,6 +65,10 @@ static void __init mpc837x_rdb_setup_arch(void)
 		mpc83xx_add_bridge(np);
 #endif
 	mpc837x_usb_cfg();
+
+#ifdef CONFIG_MMC_SDHCI_OF
+	mpc837xrdb_sd_cfg();
+#endif
 }
 
 static struct of_device_id mpc837x_ids[] = {
diff --git a/arch/powerpc/platforms/83xx/mpc83xx.h b/arch/powerpc/platforms/83xx/mpc83xx.h
index 2a7cbab..1fdf71d 100644
--- a/arch/powerpc/platforms/83xx/mpc83xx.h
+++ b/arch/powerpc/platforms/83xx/mpc83xx.h
@@ -30,6 +30,8 @@
 #define MPC8315_SICRL_USB_ULPI     0x00000054
 #define MPC837X_SICRL_USB_MASK     0xf0000000
 #define MPC837X_SICRL_USB_ULPI     0x50000000
+#define MPC837X_SICRL_USBB_MASK    0x30000000
+#define MPC837X_SICRL_SD           0x20000000
 
 /* system i/o configuration register high */
 #define MPC83XX_SICRH_OFFS         0x118
@@ -38,6 +40,8 @@
 #define MPC831X_SICRH_USB_ULPI     0x000000a0
 #define MPC8315_SICRH_USB_MASK     0x0000ff00
 #define MPC8315_SICRH_USB_ULPI     0x00000000
+#define MPC837X_SICRH_SPI_MASK     0x00000003
+#define MPC837X_SICRH_SD           0x00000001
 
 /* USB Control Register */
 #define FSL_USB2_CONTROL_OFFS      0x500
-- 
1.6.3.3

